[Dynamic RPC] Add graceful shutdown for dynamic RPC members

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75811 [RPC small change] Improve logging from 'unknown destination worker'
* **#74561 [Dynamic RPC] Add graceful shutdown for dynamic RPC members**
* #74035 [Dynamic RPC] Allow existing ranks to communicate with newly joined ranks

Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

H-Huang committed:
[Dynamic RPC] Add graceful shutdown for dynamic RPC members

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


WIP

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -395,45 +397,29 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
-        store.set(token_key, "")
-        # Other will wait for this token to be set before they execute
-        store.set(token_location, "Done")
+        # with utils.group_membership_management(store, name):
```
it's good idea to consolidate join/leave common parts into a ctx manager

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -333,9 +335,17 @@ def shutdown(graceful=True, timeout=DEFAULT_SHUTDOWN_TIMEOUT):
+                # This is a dynamic group so we need to grab the token for the operation
+                workerInfo = agent.get_worker_info()
+                with utils.group_membership_management(agent._get_store(), workerInfo.name):
+                    agent.join(shutdown=True)
```
I assume this part is WIP, as a global join won't work here?

H-Huang replied:
Yes! The join is not global anymore. Now if TPagent is part of a dynamic group, then it will leave the group when there are no active calls remaining for it.

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

TODO:
- update store allWorkerInfosKey, allWorkerInfosVector
- update group membership rpc calls

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.cpp:
```diff
@@ -95,6 +96,10 @@ std::unordered_map<std::string, worker_id_t> collectCurrentNames(
     // "Name1-Rank1,Name2-Rank2,Name3-Rank2") parse list of workers
     for (const std::string& workerInfo : splitString(allWorkerInfos, ",")) {
+      // Skip if the token is empty
+      if (workerInfo.empty()) {
```
I assume this means the entire token is empty? If thats the case, the if block should be before the for loop?

H-Huang replied:
This string was formatted in such a way that the first token of the vector of tokens was empty (due to the initial comma). This was unclear so I changed `splitString()` to `splitStringRemoveEmpties()` to also remove empty strings.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1081,6 +1082,18 @@ void TensorPipeAgent::join(bool shutdown) {
       // It is enough to wait for there to be no more active client calls, since
       // each server call corresponds to a client call for some other worker.
       callCountCV_.wait(lock, [this] { return clientActiveCalls_ == 0; });
+
+      // With dynamic RPC groups we are only concerned with active calls on
+      // local worker ActiveCallCount is 0 at this point and we will shutdown
+      // (any future calls will be dropped)
+      if (!isStaticGroup_) {
+        // Remove this agents worker info from store
+        removeCurrentName(rankToNameStore_, workerInfo_.id_, workerInfo_.name_);
+        if (shutdown) {
+          shuttingDown_ = true;
+        }
+        break;
+      }
```
Shall we move this out of the while loop or create a dedicate function for it? Because IIUC, this is no longer a `join` and this does not need a loop?

H-Huang replied:
I moved it out of the loop and into a separate function but kept it in `join` since we should keep the same API of rpc calling `.shutdown()` which joins the agent and its' thread(s) with the main process.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1235,6 +1248,24 @@ void TensorPipeAgent::updateGroupMembership(
   }
   // TODO: Rank with workerInfo is leaving, update internal mappings
```
remove todo?

mrshenli reviewed (commented):
added a few minor comments. Lint failures are real

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.cpp:
```diff
@@ -111,11 +120,9 @@ std::unordered_map<std::string, worker_id_t> collectCurrentNames(
-    allWorkerInfos = fmt::format("{}-{}", selfName, selfId);
   }
+  // Add own name to worker list
+  allWorkerInfos = fmt::format("{},{}-{}", allWorkerInfos, selfName, selfId);
```
does it make sense to add the comma to the end instead of the beginning? i.e. `fmt::format("{}{}-{},", allWorkerInfos, selfName, selfId);` or `fmt::format("{}-{},{},", selfName, selfId, allWorkerInfos);`. In this way, it seems that the parsing can be easier, i.e., we no longer need to handle the last token separately?

H-Huang replied:
This makes more sense, I updated it to `fmt::format("{}{}-{},", allWorkerInfos, selfName, selfId);` and just ignore the last comma for the string when parsing.

mrshenli reviewed (commented):
LGTM

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang commented on torch/distributed/rpc/api.py:
```diff
@@ -1,3 +1,5 @@
+__all__ = ["shutdown", "get_worker_info", "remote", "rpc_sync", "rpc_async", "RRef", "AllGatherStates", "method_factory", "new_method"]
```
This was added for compliance for new test: https://github.com/pytorch/pytorch/blob/81722f66306aef05334c27bb2a64137d8fa6493a/test/test_public_bindings.py#L284-L296

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

mrshenli reviewed (approved):
Stamp to unblock

H-Huang committed:
Update on "[Dynamic RPC] Add graceful shutdown for dynamic RPC members"


Changes:
- Refactor token implementation that was used in initialization to be a context manager that is also used in shutdown
- Graceful shutdown when the local active calls on tensorpipe agent is 0
- Expose rpc agent store instance as a readonly property.

[ghstack-poisoned]

H-Huang commented:
CI failures are unrelated

H-Huang commented:
@pytorchbot merge this

Flamefire commented:
This introduced a serious bug: https://github.com/pytorch/pytorch/issues/87359
